### DISCUSSION TITLE
I’m a little late to this dance.

### DISCUSSION THREAD

**Discussion by stitzebr** [PlummersSoftwareLLC/NightDriverStrip#405] [General]
Like I said I’m a little to this dance now trying to find a new hobby to keep my Parkinson’s brain working for a little longer. I’m a VB programmer for the last 25 years and fell in love these tiny controllers.
I’ve tried serval boards ESP-WROOM-32, and others with NightDriverStrip, a couple Ican;t use WiFi because conflicts with the display.
So I got my hands on a HelTec WiFi Dev board V3 set the PlatformIO.ini in default to demo the easiest supposably. 
All the envdefaults compile fine but when it uploads it always fails stating it’s trying ESP32 and the board is a ESP32-S3 Not ESP32 Upload error 2.
I’ve tried both demo and ledstrip getting the same error. Is there another tweak in the ini or global or secrets that doesn’t seem to be in the documentation that I’m missing? I’m just changing the ini:
[PlatformIO]
Default_envs = demo or , besides the weather api key, zip code and time zone and said and password in secrets.h

Any help guys would be great.

**Comment by robertlipe** [comment]
Hi Bret. Welcome.

You're right that there are a lot of combinations of boards. Some of them
are well tested. Some of them may be suffering from some decay. Even with
my own "off label" hardware, I wound up having to slice together parts from
platform.ini and config.h of two different boards to get the build features
and pin assignments that I needed for my exact hardware. Sounds like you
may be in a similar boat, but I think the HelTecs are somewhat popular
here. I think there was a lady just last week that was trying them, in fact.

The 'demo' build should be the easiest because there's only one external
pin involved - the data pin to the strip of LEDs. It has most features
disabled and just pounds a stream for 144 blinkies out the single pin.


So my reading is that the Heltec V3
<https://heltec.org/project/wifi-kit-32-v3/> really does have the ESP32-S3.
I think you're right that 'demo' builds for one of the first two. Lets look
at the output of  python3 tools/show_envs.py and see our choices that are
possible '-e' environments to the pio build. Of those choices, I'd predict
success with 'heltecv3demo'.

If you've been a programmer for 25 years, you know precision in a help
request matters, so please feel free to include copy-pastes of the commands
you issued, the precise errors you see, and so on.

At the beginning of that build, I see

PLATFORM: Espressif 32 (6.3.2) > Heltec WiFi Kit 32 (V3)
HARDWARE: ESP32S3 240MHz, 320KB RAM, 8MB Flash

which seems to confirm both that it's building for HeltecV3 AND gthat it's
building for ESP32S3, which is consistent. I started the build with "pio
run -v --target build  -e heltecv3demo" and it completed.
heltecv3demo              SUCCESS   00:00:04.622

If I actually had it connected to a specific serial port, I'd have tried
uploading and running like:

~/.platformio/penv/bin/pio run --target upload -*-upload-port
/dev/cu.usbserial-110 *-e heltecv3demo

On that board, I *think* the output will be on PIN5.

Good luck!
RJL

**Reply by stitzebr** [reply]
Robert, thank you for being willing to help.
Today I cloned a new copy so I'm working from scratch removing anything I may ha tried to fix it.
Here is everything done since but before building. Copied and renamed secrets.h and entered my SSID and Pass, used my real openweather API key changed to my zip and changed the time zone to America/Denver.
In the platformio.ini I added 
[platformio]
default_envs = demo

Then in a new term screen in VSCode ran the  command: 
pio run -v --target build  -e heltecv3demo

Here is the results:
Environment               Status    Duration
------------------------  --------  ------------
demo                      IGNORED
m5demo                    IGNORED
m5plusdemo                IGNORED
heltecdemo                IGNORED
heltecv2demo              IGNORED
heltecv3demo              FAILED    00:00:14.830
lilygo-tdisplay-s3-demo   IGNORED
ledstrip                  IGNORED

Do you have any suggestions?

As always thank you for helping this old man with Parkinson's.

**Reply by rbergen** [reply]
Hi @stitzebr, if I were you I'd drop the "-v --target build" from the "pio run" options. So, just execute `pio run -e heltecv3demo` and see what that says.

As your device is a Heltec V3, you can also set "default_envs" to "heltecv3demo" in platformio.ini (`default_envs = heltecv3demo`), and then just run `pio run`.

**Comment by robertlipe** [comment]
Yep, it's Pin 5.

I just added a line to the end of include/globals.h that said:
 #define LED_PIN0          godzilla

and tried to rebuild your target thusly: ~/.platformio/penv/bin/pio run -e
heltecv3demo

In file included from src/ledmatrixgfx.cpp:31:
include/globals.h:1600: warning: "LED_PIN0" redefined
 #define LED_PIN0          godzilla

In file included from src/ledmatrixgfx.cpp:31:
include/globals.h:278: note: this is the location of the previous definition
         #define LED_PIN0 5

Now I know that LED_PIN0 (the default for DEMO) is set on line 278 of
include/globals.h

    #if M5STICKC || M5STICKCPLUS || M5STACKCORE2
        #define LED_PIN0 32
    #elif LILYGOTDISPLAYS3
        #define LED_PIN0 21
    #else

*        #define LED_PIN0 5*    #endif

Untangling the interaction between platform.ini and include/globals.h isn't
always easy, so I just throw a (Godzilla-sized) wrench into it and wait for
the error to tell me where it said it saw LED_PIN0 before.

Crude? Absolutely. But lethally effective.

That concludes this edition of "teach a man to fish..."

Happy Blinking!

**Comment by robertlipe** [comment]
To help others not fall into this trap, I've just added
https://github.com/PlummersSoftwareLLC/NightDriverStrip/pull/406
to the review queue See if you think that would have helped you, @stitzebr

**Reply by stitzebr** [reply]
Yes I think so.

**Reply by robertlipe** [reply]
Excellent. Rutger integrated and accepted that change yesterday and it's live now on the front page.  When we know where we go wrong in the documentation, we try to fix it. Thanx for the feedback.

**Comment by robertlipe** [comment]
Hi, Bret.

Today I cloned a new copy so I'm working from scratch removing anything I
> may ha tried to fix it.
>
Probably wise.
 In the platformio.ini I added

> [platformio]
> default_envs = demo
>
I'd go with helltecv3demo here, just so you dont' have to keep overriding
it.



> heltecv3demo FAILED 00:00:14.830
>
 OK, so it churned away for 14 seconds before failing. Above this, there
will be the actual error message. So I planted an error, typed

 ~/.platformio/penv/bin/pio run -e heltecv3demo

..and then received:
Compiling .pio/build/heltecv3demo/src/deviceconfig.cpp.o
Compiling .pio/build/heltecv3demo/src/drawing.cpp.o
Compiling .pio/build/heltecv3demo/src/effects.cpp.o
Compiling .pio/build/heltecv3demo/src/jsonserializer.cpp.o
Compiling .pio/build/heltecv3demo/src/ledmatrixgfx.cpp.o
Compiling .pio/build/heltecv3demo/src/ledstripeffect.cpp.o
In file included from src/colordata.cpp:32:
include/globals.h:1600:2: error: #error This is an error
 #error This is an error
  ^~~~~
In file included from src/drawing.cpp:34:
include/globals.h:1600:2: error: #error This is an error
 #error This is an error

My "Error" is in include.globals.h in line 1600, where it was referenced
from drawing.cpp's line 34, where I planted a literal error message.

That's of course a little crazy, but it's just so you can see the
jibberish above that I'm trying to describe. You'll have seen 14 seconds
worth of compiler stuff (which can be a fair percentage of the project)
before yours stopped.

Where did it stop and why? You'll have a lot of "compiling X" that
succeeded and a few that didn't. Its the ones that didn't that matter and
they should sttand out.

Let's leave out the --target build for your case and just use the run
above. I added that originally becauxe I didn't want my own system to try
to run the heltec code.

> As always thank you for helping this old man with Parkinson's.
>

Glad to help, regardless of the situation. Hope you get some joy playing
with this. I have. I'm confident we'll get you to blinkage!

RJL

**Comment by stitzebr** [comment]
Gentlemen, we have blinkage! Thank you.
Any suggestions on where to start reading on getting the ledstrip to build and upload?

**Comment by robertlipe** [comment]
Excellent! Congrats!

Was there an actual error to fix or was it just One Of Those Things?

So heltecv3demo is a happy place for you, right? (This is helpful for us to
know so we can tune examples for you.)

You're ready to move on to maximum blinkage and proceed from demo
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/f42e0605ae62149521d39253909da8e81459cedc/src/effects.cpp#L333>
("just" the rainbow) to something more custom.  Look in effects.cpp (linked
above) and see which effects would look good with your combinations of
strips.

You can either choose to make a new build, supplementing #if DEMO with #if
BRET in that file to choose which effects and tweaking platform.ini to
#define BRET instead of ##defien DEMO, but being careful to keep all the
LED_PINx stuff to match your board and display and such. You can probably
copy effects from UMBRELLA or ATOM and get serious blinkage. At this point,
you'll be choosing your own effects to include. INSULATORS and XMAS_TREE
may also provide some inspiration. If you want to build up a device to
receive  effects from your computer, you can turn on the INCOMING_NETWORK
stuff and send effects from your computer.

I'll admit that I don't know the strip side of the house as well as the
hub75 (the displays) as well, so i don't have "favorites".

Just remember that 'heltecv3demo' is the name of the project in
platform.ini. It EXTENDS other projects that define the pin configuration
(read backdard that it depends on heltec-something-esel whih depends on
something that defines the pins) and it defines something like
-DDEMO that's tested in globals.h (read the "Project Configuration" block
in include/globals.h) and most importantly, src/efffects.cpp to choose the
list of effects.

Notice that the MATRIX_LENGTH is set in globals.h while the effects are in
effects.cpp. The interactions between them can be a bit non-obvous, so be
sure and read the #if DEMO blocks carefully.

You're now officially off the beaten path and on your own, but you're not
far away from existing configurations, so we should still be able to coach
you through it.

If I need to string up a batch of LEDs to walk you through it, I'm here for
you.

RJL

**Reply by stitzebr** [reply]
I think it was mostly me, I found NightDriverStrip first on Davepl/NightDriverStrip after watching his video where he says to put demo in the PlatformIO.ini and build and load. After googling the error I seen this resository and started over.
After 25 years using VB I was a bit naive to think it would really be that easy. It was you’re suggestions on the terminal commands to try is what got me here.
